From 7ce9e5e974ab574e28ef1ffc59cfe97c78efd995 Mon Sep 17 00:00:00 2001
From: Xi Ruoyao <xry111@xry111.site>
Date: Fri, 8 May 2026 13:13:37 +0800
Subject: [PATCH] calendar: Fix build failure with libical 4

Fix the following build error with libical 4 due to function signature changes:

    error: passing argument 4 of 'get_time_from_property' from incompatible pointer type [-Wincompatible-pointer-types]
    note: expected 'ICalTime * (*)(ICalProperty *)' but argument is of type 'ICalTime * (*)(const ICalProperty *)'

Link: https://github.com/libical/libical/blob/4.0/docs/MigrationGuide_to_4.md#const-pointers
---
 src/calendar-server/gnome-shell-calendar-server.c | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/src/calendar-server/gnome-shell-calendar-server.c b/src/calendar-server/gnome-shell-calendar-server.c
index 0b2f88eb7e..238969edb9 100644
--- a/src/calendar-server/gnome-shell-calendar-server.c
+++ b/src/calendar-server/gnome-shell-calendar-server.c
@@ -112,11 +112,17 @@ typedef struct
   time_t  end_time;
 } CalendarAppointment;
 
+#if ICAL_MAJOR_VERSION >= 4
+typedef ICalTime * (* get_prop_func_type) (const ICalProperty *prop);
+#else
+typedef ICalTime * (* get_prop_func_type) (ICalProperty *prop);
+#endif
+
 static gboolean
 get_time_from_property (ECalClient            *cal,
                         ICalComponent         *icomp,
                         ICalPropertyKind       prop_kind,
-                        ICalTime * (* get_prop_func) (ICalProperty *prop),
+                        get_prop_func_type     get_prop_func,
                         ICalTimezone          *default_zone,
                         ICalTime              **out_itt,
                         ICalTimezone          **out_timezone)
-- 
GitLab

